contracts/oracle/src/contract/alarms.rs

```diff
@@ -1,62 +1,216 @@
+                        above.try_into()?,
+                    )?;
+                }
+                self.price_alarms
+                    .add_alarm_below(self.storage, &self.addr, below)?;
+                Ok(Response::new())
+            }
+        }
+
+        let (below, above) = alarm.into();
+        with_quote::execute::<_, _, _, BaseC>(
+            &below,
+            AddAlarms {
+                storage,
+                addr,
+                above,
+                price_alarms: Self::PRICE_ALARMS,
+            },
+        )
+    }
+
+    fn schedule_alarm(
+        batch: &mut Batch,
+        addr: Addr,
+        next_id: &mut AlarmReplyId,
+    ) -> Result<(), ContractError> {
+        // TODO: get rid of the Nls dummy type argument
+        batch.schedule_execute_wasm_reply_always::<_, Nls>(
+            &addr,
+            ExecuteAlarmMsg::PriceAlarm(),
+            None,
+            *next_id,
+        )?;
+
+        *next_id = next_id.wrapping_add(1);
+
+        Ok(())
     }
 
-    pub fn try_notify_alarms(
+    pub fn try_notify_alarms<BaseC>(
         storage: &mut dyn Storage,
         mut batch: Batch,
         prices: &[SpotPrice],
-        max_count: u32,
-    ) -> Result<Response, ContractError> {
-        let sent = Self::PRICE_ALARMS.notify(storage, &mut batch, prices, max_count)?;
-        Ok(Response::from(batch).set_data(to_binary(&DispatchAlarmsResponse(sent))?))
+        max_count: u32, // TODO: type alias
+    ) -> Result<Response, ContractError>
+    where
+        BaseC: Currency,
+    {
+        struct NotifyCmd<'a> {
+            storage: &'a dyn Storage,
+            price_alarms: &'a PriceAlarms<'static>,
+            max_count: u32,
+            next_id: &'a mut AlarmReplyId,
+            batch: &'a mut Batch,
+        }
+
+        impl<'a, BaseC> WithQuote<BaseC> for NotifyCmd<'a>
+        where
+            BaseC: Currency,
+        {
+            type Error = ContractError;
+            type Output = u32;
+
+            fn exec<C>(self, price: Price<C, BaseC>) -> Result<Self::Output, Self::Error>
+            where
+                C: Currency,
+            {
+                let initial = *self.next_id;
+
+                self.price_alarms
+                    .alarms(self.storage, price)
+                    .take(self.max_count as usize)
+                    .try_for_each(|addr_result| {
+                        addr_result.map(|addr| {
+                            MarketAlarms::schedule_alarm(self.batch, addr, self.next_id)
+                        })?
+                    })?;
+                let processed = *self.next_id - initial;
+                Ok(processed.try_into()?)
+            }
+        }
+
+        let mut next_id = Self::MSG_ID.may_load(storage)?.unwrap_or_default();
+        let mut processed = 0u32;
+
+        prices
+            .iter()
+            .try_for_each(|price| -> Result<(), ContractError> {
+                processed += with_quote::execute::<_, _, _, BaseC>(
```

Comment:
To simplify collecting items to Batch and capping their number would suggest `NotifyCmd` return only an iterator (not sure how that would fit `WithQuote`). Then use Iterator::flatten and Iterator::take.

A not-so-good alternative would be for `NotifyCmd` to return Batch building it with Iterator::fold. Then use Batch::merge to combine into a final Batch. Here you may want to use Iterator::reduce instead of Iterator::for_each. The last piece of the puzzle is missing - how to get the length of a Batch. I suggest adding a method.